frontera/contrib/backends/remote/codecs/json.py

```diff
@@ -9,33 +9,49 @@
 from w3lib.util import to_unicode, to_bytes
 
 
-def _encode_recursively(obj):
+def _convert_and_save_type(obj):
     """
-    recursively encodes an object to unicode and store its type
+    :param obj: dict object
+
+    The purpose of this method is to transform the given dict
+    into a form that would be able to serialize with JSONEncoder.
+    In order to implement this, this method converts all byte strings
+    inside a dict to unicode and saves their type for reverting to its
+    original state. The type and the value are stored as a tuple in the
+    following format: (original_type, converted value). All other objects
+    like dict, tuple, list are converted to the same format for the sake
+    of serialization and for the ease of reverting.
+    Refer `https://github.com/scrapinghub/frontera/pull/233#discussion_r97432868`
+    for the detailed explanation about the design.
     """
     if isinstance(obj, bytes):
         return 'bytes', to_unicode(obj)
     elif isinstance(obj, dict):
-        return 'dict', [(_encode_recursively(k), _encode_recursively(v)) for k, v in six.iteritems(obj)]
+        return 'dict', [(_convert_and_save_type(k), _convert_and_save_type(v)) for k, v in six.iteritems(obj)]
     elif isinstance(obj, (list, tuple)):
-        return type(obj).__name__, [_encode_recursively(item) for item in obj]
+        return type(obj).__name__, [_convert_and_save_type(item) for item in obj]
     return 'other', obj
 
 
-def _decode_recursively(obj):
+def _revert_from_saved_type(obj):
```

Comment:
IMO revert is used in the wrong context here http://dictionary.cambridge.org/dictionary/english/revert-to-sth, convert_from_saved_type would be better.